### PR TITLE
storage: add some new APIs to the interface of Storage for storage metadata

### PR DESCRIPTION
**Pull request by c4pt0r**
### What problem does this PR solve? 

The storage engine should expose some necessary metadata to the upper layer.

### CODE REVIEW

File: util/mock/store.go [hunk 1]
```diff
@@ -49,3 +51,12 @@ func (s *Store) CurrentVersion() (kv.Version, error) { return kv.Version{}, nil
 
 // SupportDeleteRange implements kv.Storage interface.
 func (s *Store) SupportDeleteRange() bool { return false }
+
+// Name implements kv.Storage interface.
+func (s *Store) Name() string { return "mock" }
```
Comment (shenli) [inline util/mock/store.go]: Could you use a different name from `mockStorage`?

**Comment by shenli** [comment]
@c4pt0r Could you add more information in the description? Such as how will you use the new API.

**Comment by c4pt0r** [comment]
Currently, in TiDB, there are still some implementations that rely heavily on the storage engine for specific features, for example:
https://github.com/pingcap/tidb/blob/master/domain/domain.go#L585

if ebd, ok := do.store.(tikv.EtcdBackend); ok  

Like the code above, the domain should not determine the storage type  by typecasting. 

In addition, for InfoSyncer (domain/info.go), Information should not rely on specific etcd storage, which should be part of the API abstraction, and the information should be provided by the engine itself.

And here is a example, from MySQL's code:

```
typedef struct
  {
    const char *name;
    SHOW_COMP_OPTION state;
    const char *comment;
    enum db_type db_type;
    bool (*init)();
    uint slot;
    uint savepoint_offset;
    int  (*close_connection)(THD *thd);
    int  (*savepoint_set)(THD *thd, void *sv);
    int  (*savepoint_rollback)(THD *thd, void *sv);
    int  (*savepoint_release)(THD *thd, void *sv);
    int  (*commit)(THD *thd, bool all);
    int  (*rollback)(THD *thd, bool all);
    int  (*prepare)(THD *thd, bool all);
    int  (*recover)(XID *xid_list, uint len);
    int  (*commit_by_xid)(XID *xid);
    int  (*rollback_by_xid)(XID *xid);
    void *(*create_cursor_read_view)();
    void (*set_cursor_read_view)(void *);
    void (*close_cursor_read_view)(void *);
    handler *(*create)(TABLE *table);
    void (*drop_database)(char* path);
    int (*panic)(enum ha_panic_function flag);
    int (*release_temporary_latches)(THD *thd);
    int (*update_statistics)();
    int (*start_consistent_snapshot)(THD *thd);
    bool (*flush_logs)();
    bool (*show_status)(THD *thd, stat_print_fn *print, enum ha_stat_type stat);
    int (*repl_report_sent_binlog)(THD *thd, char *log_file_name, my_off_t end_offset);
    uint32 flags;
  } handlerton;
```

**Comment by c4pt0r** [comment]
will open a new PR for this.

**Pull request closed without merging**